walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,80 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    ArrayList<TimeRange> busyTimeRanges = new ArrayList();
+    ArrayList<TimeRange> freeTimeRanges = new ArrayList();
+
+    // No attendees, free the whole day
+    if(request.getAttendees().isEmpty()){
+        freeTimeRanges.add(TimeRange.WHOLE_DAY);
+        return freeTimeRanges;
+    }
+
+   
+    // Meeting last more than a day, there are other free times 
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+        return freeTimeRanges;
+    }
+
+    busyTimeRanges.addAll(getTimesRequestedAttendeesAreBusy(events, request));
+
+    if(busyTimeRanges.size() > 0){
+        Collections.sort(busyTimeRanges, TimeRange.ORDER_BY_START);
+
+        findFreeTimeBeforeFirstMeeting(busyTimeRanges, freeTimeRanges, request);
+        findFreeTimeInBetweenMeetings(busyTimeRanges, freeTimeRanges, request);
+        findFreeTimeAfterLastMeeting(busyTimeRanges, freeTimeRanges, request);
+
+    } else {
+        // Not busy, free the whole day
+        freeTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    }
+
+    return freeTimeRanges;
   }
+
+
+    private ArrayList<TimeRange> getTimesRequestedAttendeesAreBusy(Collection<Event> events, MeetingRequest request){
+        ArrayList<TimeRange> busyTimes = new ArrayList();
+        for(Event event : events){
+            // Add busy time, when attending an event
+            if(!Collections.disjoint(event.getAttendees(), request.getAttendees())){
+                busyTimes.add(event.getWhen());
+            }
+        }
+
+        return busyTimes;
+    }
+
+    private void findFreeTimeBeforeFirstMeeting(ArrayList<TimeRange> busyTimeRanges, ArrayList<TimeRange> freeTimeRanges, MeetingRequest request){
```

Comment:
For parameters, also better to use interface:  `List<TimeRange>`, which would give better flexibility.